sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -230,12 +230,15 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
   // [[func]] assumes the input is no longer null because eval already does the null check.
   @inline private[this] def buildCast[T](a: Any, func: T => Any): Any = func(a.asInstanceOf[T])
 
+  private val dateFormatter = DateFormatter()
```

Comment:
Why is this not lazy?

Reply:
It can be lazy as well but I don't see reasons for that:
- Lazy vals are not for free: http://rgg.zone/2016/12/06/scala-the-cost-of-laziness/
- `DateFormatter` is serializable and can be transferred to executors
- Instance of `DateFormatter` shouldn't consume much memory, so, we can instantiate it in the constructor.

Reply:
You are now making a dateFormatter for every cast expression out there which are mostly non-date related casts. That is also not cheap.

Reply:
I made dateFormatter lazy here